[Bug] Incorrect construction of SimpleArray from a high-dimension ndarray

I notice that when the number of the dimension > 3, the construction of SimpleArray from the ndarray would fail.

How to reproduce the bug:
```python3
import unittest
from modmesh import SimpleArrayUint64, SimpleArrayFloat64
import numpy as np

class TimeBuffer(unittest.TestCase):
    a = np.ndarray(shape= (5,5,5,5), dtype=float)
    for i in range(5):
        for j in range(5):
            for k in range(5):
                for l in range(5):
                    a[i,j,k,l] = i*1000 + j*100 + k*10 + l

    b_copy = SimpleArrayFloat64(array=a)
    print(b_copy.ndarray)
```

[Error log](https://github.com/ThreeMonth03/modmesh/actions/runs/12073106861/job/33668534857):
```console
============================= test session starts ==============================
platform linux -- Python 3.10.12, pytest-8.3.3, pluggy-1.5.0 -- /usr/bin/python3
cachedir: .pytest_cache
rootdir: /home/runner/work/modmesh/modmesh
double free or corruption (out)
Fatal Python error: Aborted

Current thread 0x00007f6be743e000 (most recent call first):
  File "/home/runner/work/modmesh/modmesh/tests/test_array.py", line 26 in TimeBuffer
  File "/home/runner/work/modmesh/modmesh/tests/test_array.py", line 5 in <module>
  File "/usr/local/lib/python3.10/dist-packages/_pytest/assertion/rewrite.py", line 184 in exec_module
  File "<frozen importlib._bootstrap>", line 688 in _load_unlocked
  File "<frozen importlib._bootstrap>", line 1006 in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 1027 in _find_and_load
  File "<frozen importlib._bootstrap>", line 1050 in _gcd_import
  File "/usr/lib/python3.10/importlib/__init__.py", line 126 in import_module
  File "/usr/local/lib/python3.10/dist-packages/_pytest/pathlib.py", line 582 in import_path
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 493 in importtestmodule
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 546 in _getobj
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 284 in obj
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 562 in _register_setup_module_fixture
  File "/usr/local/lib/python3.10/dist-packages/_pytest/python.py", line 549 in collect
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 389 in collect
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 341 in from_call
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 391 in pytest_make_collect_report
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/runner.py", line 567 in collect_one_node
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 835 in _collect_one_node
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 970 in genitems
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 975 in genitems
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 809 in perform_collect
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 347 in pytest_collection
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 336 in _main
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 283 in wrap_session
  File "/usr/local/lib/python3.10/dist-packages/_pytest/main.py", line 330 in pytest_cmdline_main
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_callers.py", line 103 in _multicall
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_manager.py", line 120 in _hookexec
  File "/usr/local/lib/python3.10/dist-packages/pluggy/_hooks.py", line 513 in __call__
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 175 in main
  File "/usr/local/lib/python3.10/dist-packages/_pytest/config/__init__.py", line 201 in console_main
  File "/usr/local/bin/pytest", line 8 in <module>

Extension modules: numpy._core._multiarray_umath, numpy.linalg._umath_linalg (total: 2)
make: *** [Makefile:[118](https://github.com/ThreeMonth03/modmesh/actions/runs/12073106861/job/33668533864#step:12:119): pytest] Aborted (core dumped)
collecting ...
```

--- a/cpp/modmesh/buffer/small_vector.hpp
+++ b/cpp/modmesh/buffer/small_vector.hpp
@@ -141,7 +141,7 @@ public:
             }
             else
             {
-                if (m_capacity < other.m_size)
+                if (m_capacity < other.m_size && m_head != m_data.data())
                 {
                     delete[] m_head;
                     m_head = nullptr;

